Add methods for finding non-global symbols by name

brenns10 opened the pull request:
This PR aims to address #47 by implementing `Prog.symbols()`, which searches for global and local symbols by a given name. Under the hood it's implemented with `drgn_program_find_symbols_by_name()` which is heavily adapted from `drgn_program_find_symbol_by_name()`, except it may return a list, and it doesn't terminate early after finding the first symbol.

I did verify this works by loading a kernel dump and trying out a few symbols. "name_show" is commonly used in sysfs, so it has several definitions. "slab_bug" is just a static function.

```python
>>> prog.symbol("name_show")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
LookupError: could not find symbol with name 'name_show'
>>> prog.symbols("name_show")
[Symbol(name='name_show', address=0xffffffff927175d0, size=0x64), Symbol(name='name_show', address=0xffffffff92b84f30, size=0x46), Symbol(name='name_show', address=0xffffffff92c1d4a0, size=0x29), Symbol(name='name_show', address=0xffffffff92d068f0, size=0x40), Symbol(name='name_show', address=0xffffffff92d17660, size=0x26), Symbol(name='name_show', address=0xffffffff92d1e3a0, size=0x23), Symbol(name='name_show', address=0xffffffff92db0da0, size=0x23), Symbol(name='name_show', address=0xffffffff92db2800, size=0x2f), Symbol(name='name_show', address=0xffffffff92db48a0, size=0x39), Symbol(name='name_show', address=0xffffffff92db4f90, size=0x29), Symbol(name='name_show', address=0xffffffff92db6ef0, size=0x23)]

>>> prog.symbol("slab_bug")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
LookupError: could not find symbol with name 'slab_bug'
>>> prog.symbols("slab_bug")
[Symbol(name='slab_bug', address=0xffffffff928a6df3, size=0xbc)]
```

For now this is more of a prototype, there are no tests and I don't know if the "symbol/symbols" naming is really good enough. It doesn't capture the distinction between global and non-global symbols. Also, there's no (implemented) way to determine whether a returned symbol is global or local, or its type or module. I don't know the best way for the API to reflect that.

I also toyed with the idea of making it so that `prog.symbols()` without any argument would simply list all symbols. This is something I've wanted to be able to do for a while, and it's quite easy with this implementation. But I'm not sure if that's a behavior you'd like to see added.

osandov commented on libdrgn/drgn.h.in:
```diff
@@ -817,6 +817,22 @@ struct drgn_error *drgn_program_find_symbol_by_name(struct drgn_program *prog,
+ * @param[out] ret Pointer to a variable to be filled with an array of symbols.
+ * The resulting value is non-@c NULL on success, and has a length of @a @c
+ * *count pointers. Each symbol sohuld be freed with @ref drgn_symbol_destroy(),
+ * and the array itself should be freed with the standard free() function.
```
Maybe we should hide the deallocation in a `void drgn_symbols_free(struct drgn_symbol **syms, size_t count)` function. Then users don't need to duplicate it and we also have the flexibility to change how the allocation works if we want.

brenns10 replied:
Adding onto this, I think it makes sense for `drgn_symbols_free()` to skip any `NULL` pointers within the `syms` array. This allows the caller to "take ownership" of any pointer by copying it and then setting its array entry to `NULL`, and I could directly use that in the Python binding rather than my static helper function.

osandov reviewed (commented):
Thanks, I've been wanting this for awhile! Other than the code nits, some thoughts on your other comments:

* I can't think of any better naming, so we should probably just elaborate on the distinction between `Program.symbol()` and `Program.symbols()` in the documentation for each.
* The binding, type, and module information should probably be added to `drgn.Symbol`. The drgn API doesn't exposes modules at all yet, so that would require additional work. But, we can add the binding and type as enums right now. We can use the `STB_*` and `STT_*` numbering from ELF for now. Would you be able to submit a separate PR for that?
* I like the idea of `prog.symbols()` with no arguments returning all symbols.

brenns10 commented on libdrgn/drgn.h.in:
```diff
@@ -817,6 +817,22 @@ struct drgn_error *drgn_program_find_symbol_by_name(struct drgn_program *prog,
+struct drgn_error *drgn_program_find_symbols_by_name(struct drgn_program *prog,
+						     const char *name,
+						     struct drgn_symbol ***ret,
+						     int *count);
```
In the implementation I preemptively set `*ret = NULL` and `*count = -1` to account for error cases, and then update them on success. Should I use `(size_t) -1` or just discard that part of the implementation, only updating the output parameters on success?

osandov replied:
Output parameters are generally undefined on error in drgn, so I'd remove those preemptive sets.

brenns10 commented on _drgn.pyi:
```diff
@@ -200,6 +201,15 @@ class Program:
+        Get a list of global and local symbols which have the given name.
+
+        :param name: The symbol name.
+        :raises LookupError: if no symbol contains the given address or matches
```
If it were a blank slate design I'd pick an empty list as a return value (since `[]` is definitely a valid `List[Symbol]`). But given that symbol() already raises an exception when it doesn't find anything, I like sticking with the exception as written. That way there are fewer surprises going between APIs.

brenns10 commented:
Second revision addresses your comments. I found and fixed a few small items (symbolp_vector_append can fail so handle that with enomem, fix the docstring for Program.symbols()). Finally, I added the change that when name=NULL, we include all symbols in the output.

I can't really find a way to compare revision 1 and 2 in the PR page, but here's a diff: https://github.com/osandov/drgn/compare/c0ebee7e175ca0be0c3ebe252cc5ff736ef4af95..1a2f4a6746ac4e206092781a58e80720f3ecbe9b

osandov commented:
I'm thinking more about what exactly I want the `symbol()`/`symbols()` interface to look like. Right now I'm thinking that we should (in this order):

1. Add `drgn.Symbol.binding` and `drgn.Symbol.type` so that we can distinguish the kinds of symbols for the following changes.
2. Make `symbol(name)` fall back to returning a local variable if no global symbols with that name are found (returning an arbitrary local symbol if there are multiple). In my opinion, that is more useful and more consistent with, e.g., `prog[name]`.
3. Merge this PR.

To be more specific about `drgn.Symbol.binding`/`drgn.Symbol.type`, I'm imagining adding:

```c
enum drgn_symbol_binding {
	DRGN_SYMBOL_BINDING_UNKNOWN,
	DRGN_SYMBOL_BINDING_LOCAL,
	DRGN_SYMBOL_BINDING_GLOBAL,
	/* ... one for each STB_* ELF constant. */
};

enum drgn_symbol_type {
	DRGN_SYMBOL_TYPE_UNKNOWN,
	DRGN_SYMBOL_TYPE_OBJECT,
	DRGN_SYMBOL_TYPE_FUNCTION,
	/* ... one for each STT_* ELF constant. */
};

enum drgn_symbol_binding drgn_symbol_binding(struct drgn_symbol *sym);

enum drgn_symbol_type drgn_symbol_type(struct drgn_symbol *sym);
```

Then we'd probably want to add a function to translate an ELF symbol to a drgn symbol and set those values in `struct drgn_symbol`.

We also need the corresponding Python bindings. `libdrgn/build-aux/gen_constants.py` can generate the Python enum wrapper for the binding/type enums.

What do you think? If you're willing to tackle these extra steps, I can give more details. Otherwise, I should be able to get to it soon.

brenns10 commented:
Sure, those all make sense. Without the `binding` and `type` fields, this PR is a bit less useful, so it makes sense to do that first. I'm happy to work on some PRs for those. Ideally I'd like to become ramped up on the C-based libdrgn components so that once I'm more comfortable, I can work on (or provide input on) some more complex stuff (such as #92 drgn.Thread API, or some sl[aou]b helpers that are comparable to the ones in crash). Doing these PRs will help me get up to speed!

osandov commented:
That sounds great! I wrote up more details in #109.

osandov commented:
Now that #109 is done, the next step is to make `Program.symbol()` fall back to returning a local symbol if no global symbols are found.

brenns10 commented:
Hi @osandov the CI checks are a bit of a moving target but I believe this is (finally) ready for another look. Since you finished up #121, I rebased this branch and made some adjustments:

- Use the `drgn_symbol_from_elf()` function which was added
- Tweaked the docstrings now that `Program.symbol()` will fall back to an arbitrary non-global symbol.
- Added tests in `test_symbol.py`, mostly that tag alongside the existing `symbol()` test, since the behavior should be similar. But  I also added one test for the `symbols()` behavior of returning all symbols.

I also added a commit to silence a lint error from black, currently on master.

osandov commented on drgn/helpers/linux/__init__.py:
```diff
@@ -33,9 +33,9 @@
 import pkgutil
 
-__all__ = []
+__all__ = []  # type: ignore
```
Even though it's somewhat magic, we might as well annotate it with its actual type.

```suggestion
__all__: List[str] = []
```

brenns10 replied:
For some reason my brain thought it was a `Dict[str, Any]` and that it wouldn't be worth annotating. Will use the List[str] instead.

osandov commented on drgn/helpers/linux/__init__.py:
```diff
@@ -33,9 +33,9 @@
+__all__ = []  # type: ignore
 for _module_info in pkgutil.iter_modules(
-    __path__,  # type: ignore[name-defined]  # python/mypy#1422
+    __path__,  # python/mypy#1422
```
It's super annoying that mypy warns for unused `type: ignore` comments which were necessary on older versions. Oh well.

The "comment on the comment" was the explanation for the `type: ignore`, so you can drop it now (and then all of the arguments can fit on one line):

```python
for _module_info in pkgutil.iter_modules(__path__, prefix=__name__ + "."):
```

osandov replied:
Oh, I guess we can add `--no-warn-unused-ignores` to the `mypy` command line. Feel free to do that instead of removing this comment, whatever you think is more useful.

brenns10 replied:
I'll add the argument so we maximize compatibility.

osandov reviewed (commented):
I'll have to look at the actual symbol stuff later, but could you please open a separate PR for the mypy thing so I can merge that right away? I left a couple of comments on that part.

brenns10 commented:
I split out the mypy fixes to #134

osandov commented:
Thanks for doing that, I'll take a look at the symbol parts either later today or tomorrow.

brenns10 commented:
(removing the mypy fix and rebasing on main)

osandov commented on _drgn.pyi:
```diff
@@ -202,13 +203,27 @@ class Program:
+        all symbols in the program.
+
+        :param name: The symbol name. If not provided, return all symbols.
+        :raises LookupError: if no symbol matches the given name
```
Now that this allows returning all symbols in the program, `LookupError` doesn't make as much sense. Maybe we should just return an empty list if nothing matches after all?

brenns10 replied:
I think we discussed it on a prior iteration, but I have no particular attachment to the LookupError. I'll return the empty list instead.

osandov replied:
Right, I think we originally decided on `LookupError` but I think the "all symbols" functionality changes the API enough that it no longer feels right.

osandov commented on libdrgn/drgn.h.in:
```diff
@@ -817,6 +817,24 @@ struct drgn_error *drgn_program_find_symbol_by_name(struct drgn_program *prog,
+ *
+ * @param[in] prog Program.
+ * @param[in] name Name to match. If @c NULL, returns all symbols.
+ * @param[out] syms_ret Pointer to a variable to be filled with an array of symbols.
+ * The resulting value is non-@c NULL on success, and has a length of
+ * @c *count_ret pointers. Each symbol shuld be freed with @ref
+ * drgn_symbol_destroy(), and the array itself should be freed with the @ref
+ * drgn_symbols_free().
+ * @param[out] count_ret Returned number of symbols in @p syms_ret.
```
We probably shouldn't mention freeing the individual symbols, because that is handled internally by `drgn_symbols_free()`. Also, let's call it `drgn_symbols_destroy()` for consistency with the rest of the libdrgn API.

```suggestion
 * @param[out] syms_ret Returned array of symbols. On success, this must be
 * freed with @ref drgn_symbols_destroy().
 * @param[out] count_ret Returned number of symbols in @p syms_ret.
```

osandov commented on libdrgn/drgn.h.in:
```diff
@@ -2622,6 +2640,14 @@ enum drgn_symbol_kind drgn_symbol_kind(struct drgn_symbol *sym);
+/**
+ * Destroy each @ref drgn_symbol in @syms, and free the array.
+ *
+ * This will ignore any @c NULL entry in the array, allowing you to take
+ * ownership of any symbol from the array prior to freeing the rest.
```
You should probably mention that symbols taken in this way must be destroyed with `drgn_symbol_destroy()`.

osandov commented on libdrgn/symbol.c:
```diff
@@ -65,3 +65,11 @@ LIBDRGN_PUBLIC bool drgn_symbol_eq(struct drgn_symbol *a, struct drgn_symbol *b)
+LIBDRGN_PUBLIC void drgn_symbols_free(struct drgn_symbol **syms, size_t count)
+{
+	for (size_t i = 0; i < count; i++)
+		if (syms[i])
+			drgn_symbol_destroy(syms[i]);
```
`drgn_symbol_destroy()` already handles a `NULL` argument.

```suggestion
		drgn_symbol_destroy(syms[i]);
```

osandov commented on tests/test_symbol.py:
```diff
@@ -135,6 +136,10 @@ def test_by_name(self):
                 self.assert_symbol_equal(prog.symbol("second"), second)
                 self.assertRaises(LookupError, prog.symbol, "third")
 
+                self.assert_symbol_equal(prog.symbols("first")[0], first)
+                self.assert_symbol_equal(prog.symbols("second")[0], second)
```
This should probably check that the list is length 1, too. Maybe add a `self.assert_symbols_equal()` method that compares a `List[drgn.Symbol]` to a `List[Symbol]` rather than hard-coding it?

osandov commented on tests/test_symbol.py:
```diff
@@ -145,9 +150,13 @@ def test_by_name_precedence(self):
         expected = 0xFFFF0008
```
You should probably add a variable for `expected - 0x8` and use that in the assertion and when creating `lower`. Maybe just `other = expected - 0x8`?

osandov commented on tests/test_symbol.py:
```diff
@@ -205,3 +218,27 @@ def test_kind(self):
+        )
+        prog = elf_symbol_program(*elf_syms)
+        symbols = prog.symbols()
+        self.assertEqual(len(symbols), 6)
+        names = Counter(sym.name for sym in symbols)
+        self.assertEqual(names["one"], 1)
+        self.assertEqual(names["two"], 2)
+        self.assertEqual(names["three"], 3)
```
The test would be more thorough if you sorted the returned symbols in some well-defined order and used the aforementioned `assert_symbols_equal`. (Also make sure to drop the `Counter` import if you do that.)

osandov commented on tests/test_symbol.py:
```diff
@@ -205,3 +218,27 @@ def test_kind(self):
+    def test_all_symbols(self):
+        elf_syms = (
+            (
+                ElfSymbol("two", 0xFFF11000, 1, STT.OBJECT, STB.LOCAL),
+                ElfSymbol("three", 0xFFF1200, 1, STT.OBJECT, STB.LOCAL),
+                ElfSymbol("one", 0xFFF10000, 1, STT.OBJECT, STB.GLOBAL),
+            ),
+            (
+                ElfSymbol("three", 0xFFF21000, 1, STT.OBJECT, STB.LOCAL),
+                ElfSymbol("two", 0xFFF20000, 1, STT.OBJECT, STB.GLOBAL),
+            ),
+            (ElfSymbol("three", 0xFFF30000, 1, STT.OBJECT, STB.GLOBAL),),
```
The reason for the test failures on the 32-bit architectures is that `elf_symbol_program()` creates a section (filled with zeroes) to contain where the symbols would be. The size is `max_address - min_address`, which in this case is 0xFFF30000 - 0xFFF1200 = 0xeff3ee00, which is greater than `sys.maxsize` (2**31-1 on 32-bit systems). Come to think of it, this is silly and I should just put these in a `NOBITS` section, but in the meantime, just assign these closer values, e.g., `0xFFFFF0000`, `0xFFFF0008`, `0xFFFF0010`, etc.

osandov replied:
6732148a11e5362f624139fe967349a1cc5e0601 changed this to not actually write the data, but I think 0xFFF1200 is a typo anyways so you might as well change these to closer values.

brenns10 replied:
Wow, thanks for this catch. This is also why my test ran for ~6 seconds rather than instantly.

brenns10 commented on _drgn.pyi:
```diff
@@ -202,13 +203,27 @@ class Program:
             the given name
         """
         ...
+    def symbols(self, name: Optional[str]) -> List[Symbol]:
```
Yes, great idea. I hadn't realized about the overlapping symbols. I can do it as part of the same C loop and callback, since the address and size are available. That would work for the Python API but the C API becomes a bit clunky. You'd either need to specify via flags what you want to search, or you'd need to provide several functions. Do you have a preference between the two APIs sketched below? (or a better alternative)
```c
/* all in one. bonus: could be extended for other search params */
enum {
    DRGN_SEARCH_NAME=(1 << 0)
    DRGN_SEARCH_ADDRESS=(1 << 1)
};
struct drgn_symbol_search {
    const char *name;
    uint64_t address;
};
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_search_symbols(struct drgn_program *prog,
			          const struct drgn_symbol_search *params,
			          unsigned int flags,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);

/* multiple C functions */
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_find_symbols_by_name(struct drgn_program *prog,
			          const char *name,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);
LIBDRGN_PUBLIC struct drgn_error *
drgn_program_find_symbols_by_address(struct drgn_program *prog,
			          uint64_t address,
				  struct drgn_symbol ***syms_ret,
				  size_t *count_ret);
```

There's not much to gain by searching for name AND address, so either works. If you think it may be useful to lookup all global symbols, or all symbols per module, or all symbols of a type (data, function, etc), then the first would be able to suit any combination of those. (not sure whether the semantic would be AND or OR)

osandov replied:
I'd prefer the separate functions for now, for consistency with `drgn_program_find_symbol_by_address` and `drgn_program_find_symbol_by_name`.

brenns10 commented on _drgn.pyi:
```diff
@@ -202,13 +203,27 @@ class Program:
         :attr:`SymbolBinding.WEAK` symbol is found, it is returned. Otherwise,
         any matching symbol (e.g., :attr:`SymbolBinding.LOCAL`) is returned. If
```
It occurs to me that these rules regarding symbol preferences seem to be only respected for name-based searches. Reading the comment of `dwfl_module_addrinfo` I don't see any documented preference in case multiple symbols match the address. Should I update that comment here? Or if you want to try to implement that preference scheme for address searches? (I could file an issue for that since it's out of scope here).

osandov replied:
It's undocumented, but `dwfl_module_addrinfo` does actually follow more or less the same rules; see the commit message for ff40f65f0db604254b47ae502cb8e3e2e2070444.

brenns10 commented:
Another round of changes here:

- No more LookupError
- I added address-based lookup. I took a "naive" approach by using the same workhorse function under the hood. I know there is an alternative: lookup the dwarf module and then iterate over just that symbol table. If you prefer I go that route, I can rewrite it. I just like the idea of using one function and callback for all searches returning multiple symbols (reduces duplication) -- even if the API doesn't expose it. Feel free to veto it!
- Added an unordered version of the assert_symbols_equal() for tests

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,135 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+		.address = address,
+		.flags = SEARCH_ADDRESS,
+	};
+	return symbols_search(prog, &arg, syms_ret, count_ret);
```
Although the code reuse is nice, it seems a bit wasteful to check every module. I think I'd rather this only check the module found with `dwfl_addrmodule`.

osandov commented on tests/test_symbol.py:
```diff
@@ -1,6 +1,7 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates.
 # SPDX-License-Identifier: GPL-3.0-or-later
 
+from collections import Counter
```
Don't need this anymore.

osandov commented on tests/test_symbol.py:
```diff
@@ -85,35 +93,69 @@ def test_by_address(self):
                 self.assert_symbol_equal(prog.symbol(0xFFFF0008), second)
+                self.assert_symbols_equal_unordered(prog.symbols(0xFFFF0008), [second])
                 self.assert_symbol_equal(prog.symbol(0xFFFF000C), second)
-                self.assertRaises(LookupError, prog.symbol, 0xFFFF0010)
```
This assertion is still valid for `prog.symbol`, right?

brenns10 commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,135 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+		return arg->err;
+	}
+
+	if (arg->results.size == 0) {
+		return drgn_error_format(DRGN_ERROR_LOOKUP,
+					 "could not find matching symbols%s",
+					 arg->bad_symtabs ?
+					 " (could not get some symbol tables)" : "");
+	}
```
I'm in agreement on silently ignoring "bad symbol table" in this case. I'll change it to simply return the empty list (NULL).

brenns10 commented:
Sorry for the delay, my other responsibilities seem to come in waves. Also sorry for the churn on the diff, I should be better at reviewing my diff before submitting.

The major adjustments in this revision:
- Don't search every module when looking up symbols by address. I still use the `symbol_search_cb()` to reuse what code I can, but instead I call it manually with the module from `dwfl_addrmodule()`.
- Don't return `DRGN_ERROR_LOOKUP` when there are no matches. One item of note -- I think that the case when the program has no debug info should still be an error, but the error was reported as `DRGN_ERROR_LOOKUP`. I changed that to `DRGN_ERROR_MISSING_DEBUG_INFO`. Maybe we should use that for the other search functions, in a second commit?
- Cleaned up the test code per your comments.
- Updated the python docstrings which I had forgotten.

brenns10 commented:
And one more tiny revision removing the `:raises: LookupError` from docstrings. Sorry!

osandov commented on _drgn.pyi:
```diff
@@ -202,13 +203,31 @@ class Program:
         ...
+    def symbols(
+        self,
+        address_or_name: Union[None, IntegerLike, str],
```
Both for type-checking and documentation, this needs the default argument:

```suggestion
        address_or_name: Union[None, IntegerLike, str] = None,
```

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+DEFINE_VECTOR(symbolp_vector, struct drgn_symbol *);
+
+enum {
+	SEARCH_NAME = (1 << 0),
+	SEARCH_ADDRESS = (1 << 1),
+	SEARCH_ALL = (1 << 2),
```
It's implied by the context, but just for extra clarity let's prefix these with `SYMBOL_`.

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+
+	symtab_len = dwfl_module_getsymtab(dwfl_module);
+	if (symtab_len == -1) {
+		arg->bad_symtabs = true;
```
This doesn't appear to be used anymore, right? Let's remove it from here and `struct symbols_search_arg`.

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+
+		name = dwfl_module_getsym_info(dwfl_module, i, &elf_sym,
+					       &elf_addr, NULL, NULL, NULL);
+		if (!symbol_match(arg, elf_addr, &elf_sym, name))
```
`dwfl_module_getsym_info` returns `NULL` on error, so we should check this here for all types of searches rather than in `symbol_match`:

```suggestion
		if (!name || !symbol_match(arg, elf_addr, &elf_sym, name))
```

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+{
+	if (arg->flags & SEARCH_ALL)
+		return true;
+	if (arg->flags & SEARCH_NAME)
+		if (name && arg->name && strcmp(name, arg->name) == 0)
+			return true;
```
It doesn't make sense to set `SEARCH_NAME` if `arg->name` is `NULL`, right? So maybe this should be (also removing the `name` check as per my other comment below):

```suggestion
	if ((arg->flags & SEARCH_NAME) && strcmp(name, arg->name) == 0)
		return true;
```

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+	if (arg->flags & SEARCH_NAME)
+		if (name && arg->name && strcmp(name, arg->name) == 0)
+			return true;
+	if (arg->flags & SEARCH_ADDRESS)
+		if (arg->address >= addr && arg->address < addr + sym->st_size)
```
Style nit: I'd squash these two if statements into one:

```suggestion
	if ((arg->flags & SEARCH_ADDRESS) &&
	    arg->address >= addr && arg->address < addr + sym->st_size)
```

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+
+		sym = malloc(sizeof(*sym));
+		if (!sym) {
+			arg->err = &drgn_enomem;
```
It seems like the only way this function can fail is with `&drgn_enomem`, so I think you can get rid of `struct symbols_search_arg::err` and have the callers always return `&drgn_enomem` if they get non-zero from the `dwfl_getmodules()` call.

osandov commented on libdrgn/program.c:
```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+		.flags = SEARCH_ADDRESS,
+	};
+
+	if (!prog->dbinfo)
+		return drgn_error_create(DRGN_ERROR_MISSING_DEBUG_INFO,
+					 "could not find matching symbols");
+
+	symbolp_vector_init(&arg.results);
+	module = dwfl_addrmodule(prog->dbinfo->dwfl, address);
+	if (module) {
+		if (symbols_search_cb(module, NULL, NULL, 0, &arg) != DWARF_CB_OK) {
+			for (size_t i = 0; i < arg.results.size; i++)
+				drgn_symbol_destroy(arg.results.data[i]);
+			symbolp_vector_deinit(&arg.results);
+			return arg.err;
+		}
+	}
+	symbolp_vector_shrink_to_fit(&arg.results);
+	*count_ret = arg.results.size;
+	*syms_ret = arg.results.data;
+	return NULL;
```
Other than how `arg` is initialized, this function is identical to `drgn_program_find_symbols_by_name()`, right? Can you refactor them to a common helper?

brenns10 replied:
They're not quite the same -- `drgn_program_find_symbols_by_name()` uses `dwfl_getmodules()` to iterate over each module and call `symbols_search_cb()` on them. Whereas this one uses `dwfl_addrmodule()` to identify the module containing the given address, and then it directly calls `symbols_search_cb()` on just that module.

I did previously have these two sharing a common helper (`symbols_search()`), but that resulted in searching every module for the case where `SYMBOL_SEARCH_ADDR` was set, and we decided to change that.

But I think you have a point here, I can re-introduce `symbols_search()` as a helper, but it can just choose whether to use `dwfl_addrmodule()` or `dwfl_getmodules()` depending on the value of `args`. That reuses most of the code, and it would allow for more general search interfaces if it was ever necessary.

osandov replied:
Duh, sorry about that, I missed that major difference which was exactly what we had just discussed. Yes, the `symbols_search()` helper you just described makes sense, so we can at least share the error handling and shrink/return code.

osandov commented on libdrgn/python/program.c:
```diff
@@ -734,6 +734,66 @@ static StackTrace *Program_stack_trace(Program *self, PyObject *args,
+static PyObject *Program_symbols(Program *self, PyObject *args, PyObject *kwargs)
+{
+	static char *keywords[] = {
+		"name", NULL,
```
Since this can be an address or name, this should probably be a positional-only argument like for `Program.symbol()`. So, change this to a `METH_O` method and document in `_drgn.pyi` that `address_or_name` is positional-only (we can't use the [PEP 570](https://www.python.org/dev/peps/pep-0570/) syntax yet since we still support Python 3.6 and 3.7).

brenns10 replied:
I chose `METH_VARARGS | METH_KWARGS` because `METH_O` does not allow for the caller to omit the argument:

```
TypeError: Program.symbols() takes exactly one argument (0 given)
```

I can however still make it positional-only by using `METH_VARARGS`, and it would probably be better since the kwargs are pointless as it is. (and the keyword argument `name` clashes with the documented parameter name in `_drgn.pyi`). So I think I'll end up switching to `METH_VARARGS`.

osandov replied:
Good point, `METH_VARARGS` makes more sense here.

osandov reviewed (changes requested):
A few minor comments, thanks!

osandov commented on libdrgn/python/program.c:
```diff
@@ -734,6 +734,66 @@ static StackTrace *Program_stack_trace(Program *self, PyObject *args,
+	size_t count;
+	PyObject *list, *arg = Py_None;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O", keywords, &arg))
```
You should include the method name here so that argument parsing exceptions include the method name:

```suggestion
	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O:symbols", keywords, &arg))
```

(This still applies to `PyArg_ParseTuple`.)

brenns10 commented:
Alright, I left open all the conversations which needed clarification, but I've resolved all of them. Feeling good about this latest iteration! I also (finally) updated the commit message to reflect the changed scope of this PR.

osandov commented:
Just did some copy editing, I'll merge this as soon as tests pass!

osandov commented:
Merged, thanks for sticking with this!
